### ISSUE TITLE
`BasicUnbufferedStreamBuf` needs to be marked for import/export

### ISSUE THREAD

**Issue by kimgr** [pocoproject/poco#921]
We have a project with a static library and an associated executable test project.

In the static library we `#include "Poco/UnbufferedStreamBuf.h"` and derive from it to implement a custom streambuf.

We then attempt to link the static library and the import library PocoFoundationd.lib into the executable, and get a linker error about multiply defined symbols:

    3>PocoFoundationd.lib(PocoFoundationd.dll) : error LNK2005: "public: __thiscall Poco::BasicUnbufferedStreamBuf<char,struct std::char_traits<char> >::BasicUnbufferedStreamBuf<char,struct std::char_traits<char> >(void)" (??0?$BasicUnbufferedStreamBuf@DU?$char_traits@D@std@@@Poco@@QAE@XZ) already defined in mylib.lib(my_custom_ostream.obj)
    3>PocoFoundationd.lib(PocoFoundationd.dll) : error LNK2005: "public: virtual __thiscall Poco::BasicUnbufferedStreamBuf<char,struct std::char_traits<char> >::~BasicUnbufferedStreamBuf<char,struct std::char_traits<char> >(void)" (??1?$BasicUnbufferedStreamBuf@DU?$char_traits@D@std@@@Poco@@UAE@XZ) already defined in mylib.lib(my_custom_ostream.obj)
    3>PocoFoundationd.lib(PocoFoundationd.dll) : error LNK2005: "protected: static int __cdecl Poco::BasicUnbufferedStreamBuf<char,struct std::char_traits<char> >::charToInt(char)" (?charToInt@?$BasicUnbufferedStreamBuf@DU?$char_traits@D@std@@@Poco@@KAHD@Z) already defined in mylib.lib(my_custom_ostream.obj)

It turns out this is because the compiler generates symbols for the inline template in the static library. The import library for PocoFoundation.dll also has its own copy of the symbol, and there's a conflict.

I was able to work around this by adding `Foundation_API` to the declaration of `BasicUnbufferedStreamBuf` in `Poco/UnbufferedStreamBuf.h`. That change makes sense to me, as all definitions of the base class are now either `dllexport` or `dllimport`, and all compilation units see the same definition.

This works on Windows, but I haven't yet checked how a Linux build behaves (does `Foundation_API` expand to anything on Unix systems?).

I'll try this as well and post back with a small patch if it appears to work. Makes sense?

**Comment by kimgr** [comment]
Also, a weird workaround is to include `Poco/Base64Decoder.h` before `Poco/UnbufferedStreamBuf.h`, because the base64 decoder class also derives from `Poco::BasicUnbufferedStreamBuf` and then marks itself as visible across module boundaries, which implicitly does the same for the base class.

That's part of the reason why this may not come up so often; if you include other Poco headers, sooner or later something happens to tag the class the right way, and it will be emitted as a weak/unresolved symbol.

**Comment by obiltschnig** [comment]
This is strange. In my knowledge class template definitions don't need  __declspec(dllexport/dllexport) declarations, so no none of the class templates defined in POCO have them.
What should work is create an explicit instantiation, and apply __declspec to that, e.g.

    template class Foundation_API BasicUnbufferedStreamBuf<char, std::char_traits<char> >; 

Could you give that a try?

**Comment by kimgr** [comment]
I agree it feels strange to add dllimport/dllexport to the template itself.

I hadn't thought about exporting the explicit specialization, that indeed fixes the build failure on Windows.

Let me try this with my various Linux toolchains as well, to make sure it doesn't break anything there, I'll try to post back before the end of the week.

Thanks!

**Comment by kimgr** [comment]
The exported explicit specialization appears to work for all our toolchains. Do you want a pull request, or could you apply your suggested fix to mainline?

Thank you!

**Comment by obiltschnig** [comment]
Also done for BufferedStreamBuf and BufferedBidirectionalStreamBuf

**Comment by obiltschnig** [comment]
Need to fix in 1.6.2 branch after it's been created.

**Comment by obiltschnig** [comment]
Unfortunately, these changes break our Travis builds for ARM targets, leading to multiple defined symbols for BufferedStreamBuf typeinfo. Thinking about that again, putting an explicit template specialization into a header file is not a good idea.

**Comment by kimgr** [comment]
No, explicit specializations in a header does feel unconventional.

Can you try my original approach in your ARM rig? Alternatively, you could try the derivation-as-typedef trick to see if you can export only UnbufferedStreamBuf;

    - typedef BasicUnbufferedStreamBuf<char, std::char_traits<char> > UnbufferedStreamBuf;
    + class Foundation_API UnbufferedStreamBuf : public BasicUnbufferedStreamBuf<char, std::char_traits<char> > {};

It's all about tricking all compilers into doing the right thing at this point.

**Comment by obiltschnig** [comment]
After some digging around, this seems to be a MSVC-only limitation. Which begs for a MSVC-only workaround ;-)

Next attempt: only provide a __declspec(dllimport) declaration of the template instantiation (wrapped in #if)

    #if defined(_WIN32) && defined(POCO_DLL) && !defined(Foundation_EXPORTS) // See #921
    template class Foundation_API BasicBufferedBidirectionalStreamBuf<char, std::char_traits<char> >;
    #endif

See if that will do.

**Comment by kimgr** [comment]
Good idea! I was so hung up on finding a portable construct, that I forgot it's only relevant for one platform :-)

I suggest the following tweaks:

- Check for _MSC_VER instead of _WIN32 -- this is an MSVC-only problem, right? Or does a mingw or Cygwin environment have the same problem? I'm not sure anymore.
- Flesh out the comment so it's manifest in the code for when/if you move off of GitHub
- s/BasicBufferedBidirectionalStreamBuf/BasicUnbufferedStreamBuf/, but I see you already edited that :-)

    #if defined(_MSC_VER) && defined(POCO_DLL) && !defined(Foundation_EXPORTS)
    // Add an exported explicit specialization so that MSVC knows not to create a copy
    // of this symbol in every object file. See issue #921 for more info.
    template class Foundation_API BasicUnbufferedStreamBuf<char, std::char_traits<char> >;
    #endif

I built my MSVC solution with this change and it worked. Since it's ifdeffed out for others, I didn't bother to test.

**Comment by obiltschnig** [comment]
fixed in poco-1.6.2 as well

**Comment by kimgr** [comment]
Thanks, much appreciated!

**Comment by gvanem** [comment]
There is still problems with this:
```
#if defined(_MSC_VER) && defined(POCO_DLL) && !defined(Foundation_EXPORTS)
// Add an exported explicit specialization so that MSVC knows not to create a copy
// of this symbol in every object file. See issue #921 for more info.
template class Foundation_API BasicUnbufferedStreamBuf<char, std::char_traits<char> >;
#endif
```

The above seems to *ass-u-me* using option `-MD` (no problem building with that). But with option `-MT`, I get the OPs link-errors too. E.g. while linking `PocoCpParser.dll`:
```
error LNK2005: "protected: __thiscall std::basic_streambuf<char,struct std::char_traits<char> >::basic_streambuf<char,struct std::char_traits<char> >(void)" (??0?$basic_streambuf@DU?$char_traits@D@std@@@std@@IAE@XZ) already defined in CppParser_Enum.obj
```

So what does a `define POCO_DLL` really mean? Using MS' dynamic runtime libs or *producing* a `PocoX.dll`? There are at least 3 cases to consider (+ the debug combos).

In `Foundation/include/Poco/Foundation.h`:
```
//
// Ensure that POCO_DLL is default unless POCO_STATIC is defined
//
#if defined(_WIN32) && defined(_DLL)
	#if !defined(POCO_DLL) && !defined(POCO_STATIC)
		#define POCO_DLL
	#endif
#endif
```
Not sure, but I feel this should be:
```
#if defined(_WIN32) && (defined(_DLL) || defined(_MT))
```
Ref.  `Foundation/include/Poco/Platform_WIN32.h`.

**Comment by zosrothko** [comment]
May be the following would satisfy everybody: Use 'extern' in all user code for the template specialization, keeping dlimport/dllexport for linking with a DLL,  except in Foudation code itself.

1/ Define Foundation_IMPEXP as
`#if (defined(_WIN32) || defined(_WIN32_WCE)) && defined(POCO_DLL)
	#if defined(Foundation_EXPORTS)
		#define Foundation_API __declspec(dllexport)
		#define Foundation_IMPEXP 
	#else
		#define Foundation_API __declspec(dllimport)	
		#define Foundation_IMPEXP extern
	#endif
#endif
`

2/ Add the specialization as BufferedStreamBuf.cpp to Foundation with content like
#include "Poco/BufferedStreamBuf.h"
namespace Poco {
Foundation_IMPEXP template class Foundation_API BasicBufferedStreamBuf<char, std::char_traits<char> >;
}

3/ Export the specialization in BufferedStreamBuf.h
Foundation_IMPEXP template class Foundation_API BasicBufferedStreamBuf<char, std::char_traits<char> >;
typedef BasicBufferedStreamBuf<char, std::char_traits<char> > BufferedStreamBuf;

By the end, there will be only one BufferedStreamBuf specialization code in the produced binary.

**Comment by zosrothko** [comment]
May be the following solution would satisfy everybody: Using 'extern' in all user code for the template specialization, keeping dlimport/dllexport for linking with a DLL, except in Foudation code itself.

1/ Define Foundation_IMPEXP as
#if (defined(_WIN32) || defined(_WIN32_WCE)) && defined(POCO_DLL)
#if defined(Foundation_EXPORTS)
#define Foundation_API __declspec(dllexport)
#define Foundation_IMPEXP
#else
#define Foundation_API __declspec(dllimport)
#define Foundation_IMPEXP extern
#endif
#endif
#ifndef Foundation_IMPEXP
	#define Foundation_IMPEXP extern
#endif

2/ Add the specialization as BufferedStreamBuf.cpp to Foundation with content like
#include "Poco/BufferedStreamBuf.h"
namespace Poco {
template class Foundation_API BasicBufferedStreamBuf >;
}

3/ Export the specialization in BufferedStreamBuf.h
Foundation_IMPEXP template class Foundation_API BasicBufferedStreamBuf >;
typedef BasicBufferedStreamBuf > BufferedStreamBuf;

By the end, there will be one and only one BufferedStreamBuf specialization code in the produced binary, even if the BufferedStreamBuf is referenced multiple times in the applicaiotn code. Moreover, the miplementation code of BufferedStreamBuf could be moved to the cpp file, if all needed specializations are defined in BufferedStreamBuf.h

**Comment by kimgr** [comment]
@gvanem: checking `_MT` is not right for determining whether `POCO_DLL` should be defined -- all it means is that you're linking with a multithreaded CRT.

It seems `POCO_DLL` means Poco itself is built as a DLL, it just happens to default to that if you're using a DLL CRT as well.

Note that your linker error is slightly different -- you're getting complaints about `std::basic_streambuf`, not Poco types. This is because you're now linking the STL statically into all DLLs, and they re-export this symbol because Poco derives from it.

Not sure how to fix this more generally.

**Comment by kimgr** [comment]
Just a heads-up -- I finally applied the patch as committed in 913f2e84b485959e433b9ee7c54c748d741ddd8e to our tree, and it failed to compile!

I know I said I tested it successfully before, so I must've screwed something up when doing so. I haven't had time to come up with a new patch, so I just threw it away. We'll probably have to find another workaround moving forward.

Sorry about the mixed messages :-/
